ee/session_recordings/ai/generate_embeddings.py

```diff
@@ -0,0 +1,192 @@
+                -- must be a completed session
+                and min_first_timestamp < now() - INTERVAL 1 DAY
+                -- let's not load all data for all time
+                -- will definitely need to do something about this length of time
+                and min_first_timestamp > now() - INTERVAL 7 DAY
+            GROUP BY session_id
+            HAVING dateDiff('second', min(min_first_timestamp), max(max_last_timestamp)) > %(min_duration_include_seconds)s
+            LIMIT %(batch_flush_size)s
+            -- when running locally the offset is used for paging
+            -- when running in celery the offset is not used
+            OFFSET %(offset)s
+        """
+
+    return [
+        x[0]
+        for x in sync_execute(
+            query,
+            {
+                "team_id": team.pk,
+                "batch_flush_size": BATCH_FLUSH_SIZE,
+                "offset": offset,
+                "min_duration_include_seconds": MIN_DURATION_INCLUDE_SECONDS,
+            },
+        )
+    ]
+
+
+def embed_batch_of_recordings(recordings: List[str], team: Team | int) -> None:
+    if isinstance(team, int):
+        team = Team.objects.get(id=team)
+
+    logger.info(f"processing {len(recordings)} recordings to embed for team {team.pk}")
+
+    while len(recordings) > 0:
+        batched_embeddings = []
+        for session_id in recordings:
+            with GENERATE_RECORDING_EMBEDDING_TIMING.time():
+                embeddings = generate_recording_embeddings(session_id=session_id, team=team)
+
+            if embeddings:
+                SESSION_EMBEDDINGS_GENERATED.inc()
+                batched_embeddings.append(
+                    {
+                        "session_id": session_id,
+                        "team_id": team.pk,
+                        "embeddings": embeddings,
+                    }
+                )
+
+        if len(batched_embeddings) > 0:
+            flush_embeddings_to_clickhouse(embeddings=batched_embeddings)
+
+
+def flush_embeddings_to_clickhouse(embeddings: List[Dict[str, Any]]) -> None:
+    sync_execute("INSERT INTO session_replay_embeddings (session_id, team_id, embeddings) VALUES", embeddings)
+    SESSION_EMBEDDINGS_WRITTEN_TO_CLICKHOUSE.inc(len(embeddings))
+
+
+def generate_recording_embeddings(session_id: str, team: Team | int) -> List[float] | None:
+    if isinstance(team, int):
+        team = Team.objects.get(id=team)
+
+    client = OpenAI()
+
+    session_metadata = SessionReplayEvents().get_metadata(session_id=str(session_id), team=team)
+    if not session_metadata:
+        logger.error(f"no session metadata found for session_id {session_id}")
+        SESSION_SKIPPED_WHEN_GENERATING_EMBEDDINGS.inc()
+        return None
+
+    session_events = SessionReplayEvents().get_events(
+        session_id=str(session_id),
+        team=team,
+        metadata=session_metadata,
+        events_to_ignore=[
+            "$feature_flag_called",
+        ],
+    )
```

Comment:
Going to have to keep an eye on performance here. It should be fine, but this is relatively expensive to go and point query each session like this from CH in a somewhat loop and schedule. You probably want to see if you can't push this query to the `OFFLINE` cluster on US at least (optimization for later)

Reply:
absolutely... I almost hope we discover this isn't useful so we don't have to productionize this process 🙈